Fixes & Improvements (previously support components)

Inrixia opened the pull request:
Fixes #24
Fixes #25 

Opening this PR to begin discussion on this feature to add support for components. This is mostly done though re https://github.com/alvgaona/cloudflare-discord-bot/pull/13 and https://github.com/alvgaona/cloudflare-discord-bot/pull/12

I'll need to update formatting first. I intend to setup a eslinf config that closely mirrors the original files format and use that since I have vs setup to properly format/fix files on saving.

Inrixia commented:
I also need to move types into the proper file and I want to check if there is a better source for the discord api types (like discord.js) as they are very incomplete

Inrixia commented:
> > I also need to move types into the proper file and I want to check if there is a better source for the discord api types (like discord.js) as they are very incomplete
> 
> @Inrixia all the types should be in here: https://discord.com/developers/docs

Yep see https://github.com/alvgaona/cloudflare-discord-bot/issues/16 I've already made good headway on switching to the official api types package, it's all uncommitted on this branch

alvgaona commented:
> > > I also need to move types into the proper file and I want to check if there is a better source for the discord api types (like discord.js) as they are very incomplete
> > 
> > 
> > @Inrixia all the types should be in here: https://discord.com/developers/docs
> 
> Yep see #16 I've already made good headway on switching to the official api types package, it's all uncommitted on this branch

Excellent. I've assigned that issue to you. Feel free to work it out.

Inrixia commented:
This has now been updated to be upstream from the latest merges to main on this repo.

This PR does two main things, adds a components field for people to add handlers for and changes behind the scenes handling for commands to use a dict for easy lookup instead of an array (which had to be searched every request).

alvgaona commented:
Hi, @Inrixia my apologies for delaying this. I promise this week I'll spend some time on it. I'll get back to you.

Inrixia commented:
@alvgaona This branch has tons of fixes including switching to new types mentioned in https://github.com/alvgaona/cloudflare-discord-bot/issues/16 and support for components. I've been updating it as I encounter issues while using it for other projects

alvgaona commented on src/permissions.ts:
```diff
@@ -1,11 +1,13 @@
-    this.types.forEach((type) => (permission += type));
+    let permission = 0n;
+    this.types.forEach((type) => (permission += PermissionFlagsBits[type]));
+    console.log(permission, String(permission));
```
@Inrixia do we want to log this?

Inrixia replied:
Oops, that's leftover from me testing. I'll remove it

alvgaona reviewed (approved):
This is looking pretty neat!

alvgaona commented:
@Inrixia feel free to merge it if you think this PR's done.

Inrixia commented:
This pr has some breaking changes afik, so maybe just double check @alvgaona

alvgaona commented:
> This pr has some breaking changes afik, so maybe just double check @alvgaona

Let me run a smoke test then.

alvgaona commented on src/handler.ts:
```diff
@@ -1,32 +1,44 @@
 
 const router = Router();
 
+export type Command<DataType extends InteractionDataType | void = void> = [
```
@Inrixia how do we create a command out of this type? Do you have an example?

Inrixia replied:
https://github.com/Inrixia/LostMokoko/blob/deb098244c4b5868632b1ad2f879a9941933a957/src/reactMentions.ts#L21

alvgaona commented on src/handler.ts:
```diff
@@ -1,32 +1,44 @@
   guildId?: string;
-  commands: [ApplicationCommand, InteractionHandler][];
+  commands: Command<any>[];
+  components?: Record<string, ComponentInteractionHandler>;
```
Any example of how to use the components in this case?

Inrixia replied:
I had a test command which returned a message with a button a user could click in this commit https://github.com/Inrixia/LostMokoko/blob/5065d55612ec136ef7360646da17be965259cd6f/src/hello.ts

Note how the `custom_id` is the same in the command definition
https://github.com/Inrixia/LostMokoko/blob/5065d55612ec136ef7360646da17be965259cd6f/src/hello.ts#L38-L44

and the component definition
https://github.com/Inrixia/LostMokoko/blob/5065d55612ec136ef7360646da17be965259cd6f/src/hello.ts#L51-L62

alvgaona commented on src/interaction.ts:
```diff
@@ -1,61 +1,88 @@
+  [InteractionDataType.ContextMenu]: CommandInteractionWithData<MenuInteractionData>;
+};
+
+export type InteractionResponse = Promise<APIInteractionResponse> | APIInteractionResponse;
```
How is this type used in terms of a client of this library?

Inrixia replied:
Assuming you are referring to `InteractionResponse` that type is the expected type any interaction handler will return.

For example this returned object conforms to the `InteractionResponse` type https://github.com/Inrixia/LostMokoko/blob/deb098244c4b5868632b1ad2f879a9941933a957/src/reactMentions.ts#L101-L106

Its specifically this response type
https://github.com/discordjs/discord-api-types/blob/35b17dcbf894b8edf0edbc4c831932192266461e/payloads/v10/_interactions/responses.ts#L48-L51

alvgaona reviewed (changes requested):
Hi @Inrixia, sorry it took me a while to look at this in depth. 

I feel like I need an example from a client's point of view. This PR touches the core of the library and we need to look at this from that point of view as well in this PR. I've tried to set up my own bot off of this PR but had some doubts about it, seems like there are missing types that we could export in order to properly set up commands for instance when writing a bot.

Do you think you could set up a "hello world" bot? It'd be very useful as we could harness that piece of code to set up testing.

Inrixia commented:
Hey no problem on the delay. Will give this a proper look over when I get time but for now you can see an example bot using this here: https://github.com/Inrixia/LostMokoko

I'll join the discord server so feel free to message me questions there or here.

Inrixia commented:
@alvgaona Have a look at my responses and the example repo. Can discuss further if you like

alvgaona commented on package.json:
```diff
@@ -29,6 +29,7 @@
 	},
 	"dependencies": {
 		"buffer": "^6.0.3",
+		"discord-api-types": "^0.33.5",
```
Let's bump this version up a little bit, to the latest, if possible.

alvgaona commented on src/interaction.ts:
```diff
@@ -1,61 +1,88 @@
-    const body = await request.json();
+import type { DictCommands } from "./handler";
+import {
+  APIMessageComponentInteraction as ComponentInteraction,
+  APIApplicationCommandInteraction as CommandInteraction,
+  APIApplicationCommandInteractionWrapper as CommandInteractionWithData,
+  APIInteractionResponse,
+  APIInteraction,
+  InteractionType,
+  APIChatInputApplicationCommandInteractionData as ChatInputData,
+  APIContextMenuInteractionData as MenuInteractionData,
```
Can you remove the `as`? I know the names are a bit longer but it's much clear.

alvgaona commented on src/interaction.ts:
```diff
@@ -1,61 +1,88 @@
 
-    const isValid = nacl.sign.detached.verify(Buffer.from(timestamp + JSON.stringify(body)), Buffer.from(signature, "hex"), Buffer.from(publicKey, "hex"));
+export enum InteractionDataType {
+  // APIChatInputApplicationCommandInteractionData
+  ChatInput,
+  // APIContextMenuInteractionData
+  ContextMenu,
```
Same here. Let's use `discord-api-types` names directly.

alvgaona commented on src/setup.ts:
```diff
@@ -1,93 +1,53 @@
 };
 
-const deleteExistingCommands = async (applicationId: string, bearer: any, guildId?: string): Promise<void> => {
+const deleteExistingCommands = async (applicationId: string, applicationSecret: string, guildId?: string): Promise<void> => {
```
```suggestion
const deleteExistingCommands = async (applicationId: string, botToken: string, guildId?: string): Promise<void> => {
```

As it's not anymore the `CLIENT_SECRET` used in here, we're using the Bot's token.

alvgaona commented on src/setup.ts:
```diff
@@ -1,93 +1,53 @@
-
-  const response = await fetch(request);
-  const commands = await response.json();
+    headers: { "Content-Type": "application/json", Authorization: `Bot ${applicationSecret}` },
```
```suggestion
    headers: { "Content-Type": "application/json", Authorization: `Bot ${token}` },
```

alvgaona commented on src/setup.ts:
```diff
@@ -1,93 +1,53 @@
-        headers: { Authorizaton: `Bearer ${bearer}` },
-      });
-    })
+        headers: { Authorization: `Bot ${applicationSecret}` },
```
```suggestion
        headers: { Authorization: `Bot ${token}` },
```

alvgaona commented on src/setup.ts:
```diff
@@ -1,93 +1,53 @@
+  guildId?: string;
+  commands: Command<any>[];
+};
+const createCommands = async ({ applicationId, guildId, commands }: createCommandsArgs, applicationSecret: string): Promise<Response> => {
```
```suggestion
const createCommands = async ({ applicationId, guildId, commands }: createCommandsArgs, token: string): Promise<Response> => {
```

alvgaona commented on src/setup.ts:
```diff
@@ -1,93 +1,53 @@
+      const response = await fetch(url, {
+        method: "POST",
+        body: JSON.stringify(command),
+        headers: { "Content-Type": "application/json", Authorization: `Bot ${applicationSecret}` },
```
```suggestion
        headers: { "Content-Type": "application/json", Authorization: `Bot ${token}` },
```

alvgaona commented on src/handler.ts:
```diff
@@ -1,32 +1,44 @@
+
 export type Application = {
   applicationId: string;
   applicationSecret: string;
```
```suggestion
  botToken: string;
```

alvgaona reviewed (changes requested):
@Inrixia finally the review. Sorry for the long wait, we're definitely improving this! Thanks a lot.

Inrixia commented:
Had a quick look at your comments. All looks good will sort this tomorrow

Inrixia commented:
Surprise, more commits.

I went ahead and did some tests and realized that this library was using Node.js libraries *naughty* for encryption so I have replaced them allowing for running code directly on workers without the need for painful transpiling.

I have also exported all the `discord-api-types` directly from the package so people won't need to go looking for them or look to install the package themselves for them.

Finally I added a proper example both in the updated readme and in the new examples folder.

Please take a look and lemme know if you need any other changes before merging

alvgaona commented on examples/hello/.gitignore:
```diff
@@ -0,0 +1,2 @@
+node_modules
+wrangler.toml
```
The file needs an empty line at the EOF.

alvgaona commented on examples/hello/package.json:
```diff
@@ -0,0 +1,12 @@
+  "main": "src/index.ts",
+  "author": "@Inrixia",
+  "dependencies": {
+    "cloudflare-discord-bot": "../../",
```
We'll change the name of the library to something nicer (follow-up PR).

alvgaona commented on examples/hello/src/hug.ts:
```diff
@@ -0,0 +1,22 @@
+  },
+  async (interaction) => {
+    const userId = interaction?.member?.user.id;
+    if (userId === undefined) throw new Error(`User is undefined`);
```
I think we can do `if (!userId)` to be a little broader. Like `""`, `null`, `undefined` would work as a falsy value.

Inrixia replied:
This is a bad idea, you always want strict checks. The typings say that you will only get undefined. Not doing strict checks can leave to unexpected behaviour.

alvgaona commented on src/setup.ts:
```diff
@@ -1,93 +1,52 @@
-  }
-
-  return `${url}/commands`;
+  if (guildId !== undefined) return RouteBases.api + Routes.applicationGuildCommands(applicationId, guildId);
+  return RouteBases.api + Routes.applicationCommands(applicationId);
```
```suggestion
  if (guildId) return RouteBases.api + Routes.applicationGuildCommands(applicationId, guildId);
  return RouteBases.api + Routes.applicationCommands(applicationId);
```

alvgaona replied:
Same idea as above.

Inrixia replied:
See above

alvgaona commented on src/interaction.ts:
```diff
@@ -1,61 +1,91 @@
+
+export enum InteractionDataType {
+  // APIChatInputApplicationCommandInteractionData
+  ChatInput,
+  // APIContextMenuInteractionData
+  ContextMenu,
```
I think we need to replace these names as they don't exist anymore in the file.

Inrixia replied:
Both names are still used?
`ChatInput` is for API**ChatInput**ApplicationCommandInteractionData
`ContextMenu` is for API**ContextMenu**InteractionData

alvgaona replied:
Yes, my bad! Don't know what I was thinking.

alvgaona commented on src/interaction.ts:
```diff
@@ -1,61 +1,91 @@
+
+  if (signature === null || timestamp === null) return false;
+
+  const encoder = new TextEncoder();
```
Neat!

alvgaona commented on src/handler.ts:
```diff
@@ -1,32 +1,50 @@
 
 const router = Router();
 
+export type Command<DataType extends InteractionDataType | void = void> = [
+  PartialAPIApplicationCommand,
+  DataType extends InteractionDataType ? CommandInteractionHandlerWithData<DataType> : CommandInteractionHandler
+];
+
```
What do you think of creating a `Component` type similar to this?

```typescript
type Component = [
 string,
 ComponentInteractionHandler
]
```

Inrixia replied:
There is no point as `ComponentInteractionHandler` is already exported to the user. Also the type you specified here is wrong, Components use a dict not an array.

alvgaona commented on src/handler.ts:
```diff
@@ -1,32 +1,50 @@
   publicKey: string;
   guildId?: string;
-  commands: [ApplicationCommand, InteractionHandler][];
+  commands: Command<any>[];
```
Should we change `any` to a more restrictive type? WDYT?

Inrixia replied:
I always avoid using `any` where possible but in this case narrowing is not possible as different command signatures are incompatible with each other.

alvgaona reviewed (changes requested):
Nice changes @Inrixia! Left a few comments, nothing to big and a few questions for you.

Regarding the example: can we move away the example to a different repository? If you'd like to have the example linked to this repo we could use `git submodules`. 

I have created this https://github.com/discordcf/hello-bot to make it public once we make a new release of the framework.

Inrixia commented:
@alvgaona great comments. Will update when I get home

Inrixia commented:
> Regarding the example: can we move away the example to a different repository? If you'd like to have the example linked to this repo we could use `git submodules`. Additionally, we could automate the update every time a push event is emitted from the main branch in that repo (but this is an optimization).

This is over engineering, having multiple repos for examples etc is more difficult all round especially for users to find them (most of the time they won't) and makes maintaining them harder. Also the fact that you suggest adding automation to update the main repo when changes are pushed from the secondary is a pretty clear indicator that it belongs in the main repo.

Speaking from experience of converting a major project to submodules in a github org (at least 8 repos + 2 helper repos and 1 "main" repo) and back again submodules look nice but can be a nightmare in practise, they have their use but this is definitely not it and I personally would advocate against using them as monorepos are much much easier to deal with.

alvgaona reviewed (changes requested):
Just reopened a conversation. Let's add that, I'm afraid that for some reason Discord API changes for some reason and breaks us.

alvgaona commented:
> > Regarding the example: can we move away the example to a different repository? If you'd like to have the example linked to this repo we could use `git submodules`. Additionally, we could automate the update every time a push event is emitted from the main branch in that repo (but this is an optimization).
> 
> This is over engineering, having multiple repos for examples etc is more difficult all round especially for users to find them (most of the time they won't) and makes maintaining them harder. Also the fact that you suggest adding automation to update the main repo when changes are pushed from the secondary is a pretty clear indicator that it belongs in the main repo.
> 
> Speaking from experience of converting a major project to submodules in a github org (at least 8 repos + 2 helper repos and 1 "main" repo) and back again submodules look nice but can be a nightmare in practise, they have their use but this is definitely not it and I personally would advocate against using them as monorepos are much much easier to deal with.

@Inrixia I'm not suggesting we use a repo for each example bot. At the moment this is the only example we've got but we could do a monorepo for the examples, I'm in favor of doing that. I would prefer we keep this core repo as lean as possible, and just link to the examples repo in the documentation. If we don't want to do submodules, I'm fine with it; although, I do insist we keep the examples in a repo named `discordcf/example-bots` or sth.

Inrixia commented:
@alvgaona If we don't have the examples embedded in the repo then submodules is a good intermediary as at least on github it links nicely on the web ui.

Inrixia commented:
Also you should rename all the example repos to be prefixed with `examples-` or some other tag so it's easier to find them, currently it's not obvious at all what they are.
